src/textual/worker.py

```diff
@@ -0,0 +1,388 @@
+from __future__ import annotations
+
+import asyncio
+import enum
+import inspect
+from contextvars import ContextVar
+from time import monotonic
+from typing import (
+    TYPE_CHECKING,
+    Awaitable,
+    Callable,
+    Coroutine,
+    Generic,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import rich.repr
+from rich.traceback import Traceback
+from typing_extensions import TypeAlias
+
+from .message import Message
+
+if TYPE_CHECKING:
+    from .app import App
+    from .dom import DOMNode
+
+
+active_worker: ContextVar[Worker] = ContextVar("active_worker")
+"""Currently active worker context var."""
+
+
+class NoActiveWorker(Exception):
+    """There is no active worker."""
+
+
+class WorkerError(Exception):
+    """A worker related error."""
+
+
+class WorkerFailed(WorkerError):
+    """The worker raised an exception and did not complete."""
+
+    def __init__(self, error: BaseException) -> None:
+        self.error = error
+        super().__init__(f"Worker raise exception: {error!r}")
+
+
+class DeadlockError(WorkerError):
+    """The operation would result in a deadlock."""
+
+
+class WorkerCancelled(WorkerError):
+    """The worker was cancelled and did not complete."""
+
+
+def get_current_worker() -> Worker:
+    """Get the currently active worker.
+
+    Raises:
+        NoActiveWorker: If there is no active worker.
+
+    Returns:
+        A Worker instance.
+    """
+    try:
+        return active_worker.get()
+    except LookupError:
+        raise NoActiveWorker("There is no active worker in this task or thread.")
```

Comment:
Wouldn't we want to suppress the part of the error that is related to the `LookupError`?

```suggestion
        raise NoActiveWorker("There is no active worker in this task or thread.") from None
```